src/main/java/org/breedinginsight/api/auth/rules/ExperimentCollaboratorSecuredAnnotationRule.java

```diff
@@ -0,0 +1,156 @@
+    @Inject
+    private BrAPITrialDAO brAPITrialDAO;
+    @Inject
+    private ExperimentalCollaboratorDAO experimentalCollaboratorDAO;
+
+    @Override
+    public SecurityRuleResult check(HttpRequest<?> request, @Nullable RouteMatch<?> routeMatch, @Nullable Map<String, Object> claims) {
+        // Does not approve request so that checks after it can check. Only rejects on fail.
+
+        if (routeMatch instanceof MethodBasedRouteMatch) {
+            MethodBasedRouteMatch methodRoute = ((MethodBasedRouteMatch) routeMatch);
+
+            if (methodRoute.hasAnnotation(ExperimentCollaboratorSecured.class)) {
+                String programId = (String) routeMatch.getVariableValues()
+                        .get("programId");
+                String experimentId = extractExperimentId(routeMatch);
+                if (programId == null) {
+                    throw new HttpServerException("Endpoint does not have program id to check roles against");
+                }
+                if (experimentId == null) {
+                    throw new HttpServerException("Endpoint does not have experiment id to check roles against");
+                }
+
+                if (!programDAO.existsById(UUID.fromString(programId))) {
+                    throw new HttpStatusException(HttpStatus.NOT_FOUND, "Program does not exist");
+                }
+                Optional<BrAPITrial> trial;
+                try {
+                    trial = brAPITrialDAO.getTrialById(UUID.fromString(programId), UUID.fromString(experimentId));
+                } catch (ApiException e) {
+                    throw new RuntimeException(e);
+                } catch (DoesNotExistException e) {
+                    throw new HttpStatusException(HttpStatus.NOT_FOUND, "Experiment does not exist");
+                }
+                if( trial.isEmpty()) {
+                    throw new HttpStatusException(HttpStatus.NOT_FOUND, "Experiment does not exist");
+                }
+
+                if (claims != null){
+                    AuthenticatedUser user = securityService.getUser();
+
+                    return processExperiment(user, experimentId, programId);
+                }
+
+                // Rejects if no claims
+                return SecurityRuleResult.REJECTED;
+            }
+        }
+
+        return SecurityRuleResult.UNKNOWN;
+    }
+
+    private static String extractExperimentId(@NotNull RouteMatch<?> routeMatch) {
+        //The endpoints can use either the "experimentId" or "trialId" parameter to pass the experiment ID
+        String experimentId = (String) routeMatch.getVariableValues()
+                .get("experimentId");
+        if( experimentId==null) {
+            experimentId = (String) routeMatch.getVariableValues()
+                    .get("trialId");
+        }
+        return experimentId;
+    }
+
+    private SecurityRuleResult processExperiment(AuthenticatedUser authenticatedUser, String experimentId, String programId) {
```

Comment:
Could you rename this method? We use the term "process" in a lot of our import code, and I think for that reason it is confusing to use it here. Something like `checkAuthorization` or `getSecurityRuleResult` would be more clear.

Reply:
renamed ` checkAuthorization`